### ISSUE TITLE
Getting the transformation parameters

### ISSUE THREAD

**Issue by caro-andonie** [voxelmorph/voxelmorph#371]
**Details of experiments**

There is a way to get the transformation parameters between the fixed and moving images? In particular rotation and translation.

**Comment by thomasediolausson** [comment]
I don't know if this could help you, but I did something like this for my own project using the code in this repository

What I needed were rigid body transformation parameters (i.e. translations and rotations) for registering two images. I used the [ParamsToAffineMatrix](https://github.com/voxelmorph/voxelmorph/blob/f14939a6db68e14b0e1b1fefe4b97ce90852809e/voxelmorph/tf/layers.py#L447) layer to convert a dense layer to an affine matrix. I noticed that it expects inputs of translations first, then rotations. So a 45 degree rotation on the y-z plane would be something like 0, 0, 0, 45, 0, 0. 

Then I sent this to the [AffineToDenseShift](https://github.com/voxelmorph/voxelmorph/blob/f14939a6db68e14b0e1b1fefe4b97ce90852809e/voxelmorph/tf/layers.py#L502) layer. The output of this works together with the [SpatialTransformer](https://github.com/voxelmorph/voxelmorph/blob/f14939a6db68e14b0e1b1fefe4b97ce90852809e/voxelmorph/tf/layers.py#L34) layer. 

So in the output of my model, I received the transformed image from the SpatialTransformer layer, and the six parameters (translations and rotations) that are sent to ParamsToAffineMatrix layer.

I hope that this made sense, and that it could help you with your experiment.

**Comment by dfvr1994** [comment]
Hi @thomasediolausson , mind sharing the code used to such approach? Thanks

**Comment by thomasediolausson** [comment]
Hi @dfvr1994 , the experiments I made are very messy and I can't share them entirely due to privacy reasons. It was a while ago I did this, so I haven't tested it since.

I hope this helps! 
I had a standard keras model function:
````

def my_model():
...
    rigid_body_transformation = layers.Dense(units=6, activation=None, name="Parameters")(last_layer_of_network)
    affine_matrix = ParamsToAffineMatrix()(rigid_body_transformation)
    rigid_transformed_volume = SpatialTransformer(name="Transformed")([moving,affine_matrix])
    model = keras.Model(inputs=[reference,moving],outputs=[rigid_transformed_volume,rigid_body_transformation], name="CNN_Model")
...
    model.compile(optimizer=tf.keras.optimizers.SGD(
                learning_rate=learning_rate, 
                momentum=0.0, 
                nesterov=False, 
                name="SGD"),
                loss={
                    "Transformed": NCC(win=9).loss
                },
                loss_weights={
                    "Transformed": 1
                }
                )

    return model

**Comment by adalca** [comment]
I'm sorry we missed this conversation earlier -- are you trying to predict an affine/rigid matrix directly, or want to get a deformable deformation field and fit affine/rigid parameters to that?

**Comment by dfvr1994** [comment]
No worries @adalca -- In my case I want to predict the affine/rigid matrix directly. Any suggestion on how to achieve this?

**Comment by adalca** [comment]
@mu40 has played with this quite a bit, but the simplest is to have a Dense layer attached to the UNet (or just the "encoding" part of the UNet) just like  @thomasediolausson  says. Did you try that?
